src/line.js

```diff
@@ -44,22 +51,91 @@ define([
   };
 
   Line.prototype.addSourceGroup = function(sourceGroup) {
-    this._sources.push(sourceGroup);
+    var newSource = {
+      sourceGroup: sourceGroup,
+      prevSourceId: null,
+      nextSourceId: null
+    };
+
+    var source = sourceGroup.getSource();
+
+    if (this._firstSourceId) {
+      var currentSource = null;
+
+      do {
+        if (!currentSource) {
+          currentSource = this._sources[this._firstSourceId];
+        }
+        else {
+          currentSource = this._sources[currentSource.nextSourceId];
+        }
+
+        if (source.endTime < currentSource.sourceGroup.getSource().startTime) {
+          if (currentSource.prevSourceId) {
+            if (source.startTime < this._sources[currentSource.prevSourceId]
+                                      .sourceGroup.getSource().endTime) {
+              throw new RangeError('peaks.sources-group._addsource(): sources must not overlap');
+            }
+
+            this._sources[currentSource.prevSourceId].nextSourceId = source.id;
+            newSource.prevSourceId = currentSource.prevSourceId;
+          }
+          else {
+            this._firstSourceId = source.id;
+          }
+
+          currentSource.prevSourceId = source.id;
+          newSource.nextSourceId = currentSource.sourceGroup.getSource().id;
+
+          this._sources[source.id] = newSource;
+          break;
+        }
+      } while (currentSource.nextSourceId);
+
+      if (!newSource.prevSourceId && !newSource.nextSourceId) {
+        currentSource.nextSourceId = source.id;
+        newSource.prevSourceId = currentSource.sourceGroup.getSource().id;
+        this._sources[source.id] = newSource;
+      }
+    }
+    else {
+      this._firstSourceId = source.id;
+      this._sources[source.id] = newSource;
+    }
+
     sourceGroup.addToGroup(this._group);
   };
 
   Line.prototype.removeSourceGroup = function(sourceGroup) {
-    var index = this._sources.indexOf(sourceGroup);
+    var source = sourceGroup.getSource();
+
+    if (this._sources[source.id].prevSourceId) {
+      this._sources[this._sources[source.id].prevSourceId].nextSourceId
```

Comment:
it's tricky to read, but for me, it looks like removing a node from a linked list.

Reply:
Yes, it is a linked list. We don't have this structure natively right? I should implement a proper LinkedList maybe?